classes/Deck.js

```diff
@@ -1,22 +1,115 @@
-import { PlayingCard } from './PlayingCard'
+import { PlayingCard, SUITES, RANKS } from './playing-card'
 
 class Deck {
-  constructor(suites, ranks) {
+  constructor(cards) {
+    this.cards = cards
+  }
+
+  static from(suites, ranks) {
     const cards = []
+    let i = 0
     for (const suite of suites) {
       for (const rank of ranks) {
-        cards.push(new PlayingCard(suite, rank))
+        cards.push(new PlayingCard(suite, rank, i++))
       }
     }
-    this.cards = cards
+    return cards
+  }
+
+  print() {
+    for (const card of this.cards) {
+      console.log(card.toString())
+    }
+  }
+
+  drawTopCard() {
+    this.validateNotEmpty('drawTopCard')
+
+    const card = this.cards.pop()
+    this.checkIsEmpty()
+
+    return card
+  }
+  
+  drawBottomCard() {
+    this.validateNotEmpty('drawBottomCard')
+
+    const card = this.cards.shift()
+    this.checkIsEmpty()
+
+    return card
+  }
+
+  insertCardTop(card) {
+    this.cards.push(card)
+  }
+
+  insertCardBottom(card) {
+    this.cards.unshift(card)
+  }
+
+  checkIsEmpty() {
+    this.isEmpty = this.cards.length === 0
+  }
```

Comment:
The name of this method does not reflect what's going on here. You are just making an assignment, not checking anything.

Reply:
Also, this variable is completely useless. Just call this method anytime you want to figure out if the deck is empty and return this.cards.length === 0. Calculating the length each time you want to figure out if it's empty or not will most likely not have any significant performance problems, and keeping more state is more difficult to maintain than less state. If you can derive the state from other state, it's usually better than keeping new state.